pkg/kv/kvclient/kvcoord/txncorrectnesstest/txncorrectnesstest_test: TestTxnDBReadSkewAnomaly failed

pkg/kv/kvclient/kvcoord/txncorrectnesstest/txncorrectnesstest_test.TestTxnDBReadSkewAnomaly [failed](https://mesolite.cluster.engflow.com/invocations/default/7eb10f91-9136-4256-bf70-a316fad5e4d3?testReportRun=1&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZjbGllbnQva3Zjb29yZC90eG5jb3JyZWN0bmVzc3Rlc3Q6dHhuY29ycmVjdG5lc3N0ZXN0X3Rlc3Q=) on master @ [18eb14ffe6fd73a57de81ccf3a9dd415398f51c4](https://github.com/cockroachdb/cockroach/commits/18eb14ffe6fd73a57de81ccf3a9dd415398f51c4):

Fatal error:

```
panic: no stores for meansForStoreSet: []

```
Stack: 

```
goroutine 37898 [running]:
github.com/cockroachdb/cockroach/pkg/kv/kvserver/allocator/mmaprototype.computeMeansForStoreSet({0x9eba8e0, 0xc004a1da40}, {0x0, 0x0, 0x0}, 0xc001fab4d0, 0xc001fab500)
	pkg/kv/kvserver/allocator/mmaprototype/load.go:483 +0x6ce
github.com/cockroachdb/cockroach/pkg/kv/kvserver/allocator/mmaprototype.(*meansMemo).getMeans(0xc001fab4a0, {0x0, 0x0, 0x0})
	pkg/kv/kvserver/allocator/mmaprototype/load.go:448 +0x259
github.com/cockroachdb/cockroach/pkg/kv/kvserver/allocator/mmaprototype.(*clusterState).processStoreLeaseholderMsgInternal(0xc004a1da40, {0x9ec9da8, 0xc0021a3c20}, 0xc001c9f000, 0x80, 0xc00122e3f0)
	pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go:1759 +0x94f
github.com/cockroachdb/cockroach/pkg/kv/kvserver/allocator/mmaprototype.(*clusterState).processStoreLeaseholderMsg(...)
	pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go:1476
github.com/cockroachdb/cockroach/pkg/kv/kvserver/allocator/mmaprototype.(*allocatorState).ComputeChanges(0xc004487260, {0x9ec9da8, 0xc0021a3c20}, 0xc001c9f000, {0x21a3c20?, 0xc0?, 0x0?})
	pkg/kv/kvserver/allocator/mmaprototype/allocator_state.go:349 +0x19b
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*mmaStoreRebalancer).rebalance(0xc001b6aa80, {0x9ec9da8, 0xc0021a3c20}, 0x1)
	pkg/kv/kvserver/mma_store_rebalancer.go:158 +0x368
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*mmaStoreRebalancer).run(0xc001b6aa80, {0x9ec9da8, 0xc0021a3c20}, 0xc004a1b7c0)
	pkg/kv/kvserver/mma_store_rebalancer.go:109 +0x490
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*mmaStoreRebalancer).start.func1({0x9ec9da8, 0xc0021a3c20})
	pkg/kv/kvserver/mma_store_rebalancer.go:138 +0x4d
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func1({0x9ec9da8, 0xc0021a3c20})
	pkg/util/stop/stopper.go:447 +0x113
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 37820
	pkg/util/stop/stopper.go:445 +0x247
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestTxnDBReadSkewAnomaly
    test_log_scope.go:172: test logs captured to: outputs.zip/logTestTxnDBReadSkewAnomaly1540145003
    test_log_scope.go:82: use -show-logs to present logs inline
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTxnDBReadSkewAnomaly.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-63857

--- a/pkg/kv/kvserver/allocator/mmaprototype/allocator_state.go
+++ b/pkg/kv/kvserver/allocator/mmaprototype/allocator_state.go
@@ -959,7 +959,12 @@ func (cs *clusterState) computeCandidatesForReplicaTransfer(
 	// these stores before computing the means, we can use it verbatim, otherwise
 	// we will recompute the means again below.
 	cs.scratchDisj[0] = conj
-	means := cs.meansMemo.getMeans(cs.scratchDisj[:1])
+	means, ok := cs.meansMemo.getMeans(cs.scratchDisj[:1])
+	if !ok {
+		// No store satisfies the constraint expression.
+		passObs.replicaShed(ctx, noCandidate)
+		return candidateSet{}, sheddingSLS
+	}
 
 	// Pre-means filtering: copy to scratch, then filter in place.
 	// Filter out stores that have a non-OK replica disposition.
@@ -979,8 +984,16 @@ func (cs *clusterState) computeCandidatesForReplicaTransfer(
 		return candidateSet{}, sheddingSLS
 	} else {
 		// Some stores were filtered; recompute means over filtered set.
-		cs.scratchMeans = computeMeansForStoreSet(
+		// filteredStores is non-empty (checked above).
+		var ok bool
+		cs.scratchMeans, ok = computeMeansForStoreSet(
 			cs, filteredStores, cs.meansMemo.scratchNodes, cs.meansMemo.scratchStores)
+		if !ok {
+			// Unreachable: filteredStores is non-empty (checked above). Gate
+			// the assert so variadic arg boxing doesn't allocate on the
+			// success path.
+			assertTruef(ctx, false, "computeMeansForStoreSet returned !ok for non-empty filteredStores=%v", filteredStores)
+		}
 		effectiveMeans = &cs.scratchMeans
 		ml.logf(ctx, 3,
 			"pre-means filtered %d stores → remaining %v, means: store=%v node=%v",

--- a/pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go
+++ b/pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go
@@ -1756,7 +1756,12 @@ func (cs *clusterState) processStoreLeaseholderMsgInternal(
 	}
 	localss := cs.stores[msg.StoreID]
 	cs.meansMemo.clear()
-	clusterMeans := cs.meansMemo.getMeans(nil)
+	clusterMeans, ok := cs.meansMemo.getMeans(nil)
+	if !ok {
+		// No stores are known to the allocator yet (e.g. the local store
+		// descriptor has not been gossiped at startup). Nothing to do.
+		return
+	}
 	// Sort by store ID for deterministic log output in tests.
 	storeIDs := make([]roachpb.StoreID, 0, len(cs.stores))
 	for storeID := range cs.stores {
@@ -2655,7 +2660,10 @@ func (cs *clusterState) computeLoadSummary(
 // TODO(wenyihu6): check to make sure obs here is correct
 func (cs *clusterState) loadSummaryForAllStores(ctx context.Context) string {
 	var buf redact.StringBuilder
-	clusterMeans := cs.meansMemo.getMeans(nil)
+	clusterMeans, ok := cs.meansMemo.getMeans(nil)
+	if !ok {
+		return "no stores known to allocator"
+	}
 	buf.Printf("cluster means: (stores-load %v) (stores-capacity %v)\n",
 		clusterMeans.storeLoad.load, clusterMeans.storeLoad.capacity)
 	buf.Printf("(nodes-cpu-load %v) (nodes-cpu-capacity %v)\n",

--- a/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go
+++ b/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go
@@ -217,7 +217,11 @@ func (re *rebalanceEnv) rebalanceStores(
 	// responsible for equalizing load across two nodes that have 30% and 50%
 	// cpu utilization while the cluster mean is 70% utilization (as an
 	// example).
-	clusterMeans := re.meansMemo.getMeans(nil)
+	clusterMeans, ok := re.meansMemo.getMeans(nil)
+	if !ok {
+		passML.logf(ctx, 2, "no stores known to allocator; skipping rebalancing pass")
+		return nil
+	}
 	var sheddingStores []sheddingStore
 	passML.logf(ctx, 2,
 		"cluster means: (stores-load %s) (stores-capacity %s) (nodes-cpu-load %d) (nodes-cpu-capacity %d)",
@@ -897,7 +901,13 @@ func (re *rebalanceEnv) rebalanceLeasesFromLocalStoreID(
 
 		// NB: candsPL is not empty - it includes at least the current leaseholder
 		// and one additional candidate.
-		means := computeMeansForStoreSet(re, candsPL, scratchNodes, scratchStores)
+		means, ok := computeMeansForStoreSet(re, candsPL, scratchNodes, scratchStores)
+		if !ok {
+			// Unreachable: candsPL is non-empty. Gate the assert so variadic
+			// arg boxing doesn't allocate on the success path (this site is in
+			// a hot inner loop).
+			assertTruef(ctx, false, "computeMeansForStoreSet returned !ok for non-empty candsPL=%v", candsPL)
+		}
 		sls := re.computeLoadSummary(ctx, store.StoreID, &means.storeLoad, &means.nodeLoad, ml)
 		if sls.dimSummary[CPURate] < overloadSlow {
 			// This store is not cpu overloaded relative to these candidates for

--- a/pkg/kv/kvserver/allocator/mmaprototype/load.go
+++ b/pkg/kv/kvserver/allocator/mmaprototype/load.go
@@ -327,6 +327,11 @@ func (m meanNodeLoad) SafeFormat(w redact.SafePrinter, _ rune) {
 // storeLoadSummary for some of the stores in this set. This cache is lazily
 // populated and is implicitly invalidated if the loadSeqNum of the
 // storeLoadSummary no longer matches that of storeState.loadSeqNum.
+//
+// An entry with len(stores) == 0 represents a constraint expression that
+// no store satisfies. Such entries are still cached (for O(1) repeat
+// lookups) but meansLoad is left zero; meansMemo.getMeans signals this
+// case to callers via its ok return.
 type meansForStoreSet struct {
 	constraintsDisj
 	meansLoad
@@ -437,16 +442,24 @@ type loadInfoProvider interface {
 	computeLoadSummary(context.Context, roachpb.StoreID, *meanStoreLoad, *meanNodeLoad, mmaLogger) storeLoadSummary
 }
 
-// getMeans returns the means for an expression.
-func (mm *meansMemo) getMeans(expr constraintsDisj) *meansForStoreSet {
-	means, ok := mm.meansMap.get(expr)
-	if ok {
-		return means
+// getMeans returns the means for an expression. Returns ok=false if no
+// stores satisfy the expression (or, for a nil expr, if the allocator knows
+// of no stores at all). When ok is false the returned *meansForStoreSet
+// value is unspecified; callers must not pass it to getStoreLoadSummary or
+// otherwise consume it. Empty results are still inserted into the cache so
+// repeated lookups within an allocator pass remain O(1).
+func (mm *meansMemo) getMeans(expr constraintsDisj) (*meansForStoreSet, bool) {
+	means, hit := mm.meansMap.get(expr)
+	if hit {
+		return means, len(means.stores) > 0
 	}
 	means.constraintsDisj = expr
 	mm.constraintMatcher.constrainStoresForExpr(expr, &means.stores)
-	means.meansLoad = computeMeansForStoreSet(mm.loadInfoProvider, means.stores, mm.scratchNodes, mm.scratchStores)
-	return means
+	if len(means.stores) == 0 {
+		return means, false
+	}
+	means.meansLoad, _ = computeMeansForStoreSet(mm.loadInfoProvider, means.stores, mm.scratchNodes, mm.scratchStores)
+	return means, true
 }
 
 // getStoreLoadSummary returns the load summary for a store in the context of
@@ -471,6 +484,11 @@ func (mm *meansMemo) getStoreLoadSummary(
 // stores may contain duplicate storeIDs, in which case computeMeansForStoreSet
 // should deduplicate processing of the stores. stores should be immutable.
 //
+// If stores is empty, computeMeansForStoreSet returns the zero meansLoad
+// and ok=false. Callers must check ok before consuming the returned means;
+// the zero value would otherwise misclassify stores as overloaded due to
+// zero capacity.
+//
 // TODO: fix callers to exclude stores based on node failure detection, from
 // the mean.
 func computeMeansForStoreSet(
@@ -478,9 +496,9 @@ func computeMeansForStoreSet(
 	stores []roachpb.StoreID,
 	scratchNodes map[roachpb.NodeID]*NodeLoad,
 	scratchStores map[roachpb.StoreID]struct{},
-) (means meansLoad) {
+) (means meansLoad, ok bool) {
 	if len(stores) == 0 {
-		panic(fmt.Sprintf("no stores for meansForStoreSet: %v", stores))
+		return meansLoad{}, false
 	}
 	clear(scratchNodes)
 	clear(scratchStores)
@@ -540,7 +558,7 @@ func computeMeansForStoreSet(
 		float64(means.nodeLoad.loadCPU) / float64(means.nodeLoad.capacityCPU)
 	means.nodeLoad.loadCPU /= LoadValue(n)
 	means.nodeLoad.capacityCPU /= LoadValue(n)
-	return means
+	return means, true
 }
 
 // loadSummary aggregates across all load dimensions for a store, or a node.

--- a/pkg/kv/kvserver/allocator/mmaprototype/load_test.go
+++ b/pkg/kv/kvserver/allocator/mmaprototype/load_test.go
@@ -312,7 +312,11 @@ func TestMeansMemo(t *testing.T) {
 					cc := parseConstraints(t, parts)
 					disj = append(disj, interner.internConstraintsConj(cc))
 				}
-				mss = mm.getMeans(disj)
+				var ok bool
+				mss, ok = mm.getMeans(disj)
+				if !ok {
+					return "ok: false (no stores match)\n"
+				}
 				var b strings.Builder
 				fmt.Fprintf(&b, "stores: ")
 				printPostingList(&b, mss.stores)
@@ -358,4 +362,30 @@ func TestMeansMemo(t *testing.T) {
 				return fmt.Sprintf("unknown command: %s", d.Cmd)
 			}
 		})
+}
+
+// TestComputeMeansForStoreSetEmpty verifies that computeMeansForStoreSet
+// returns ok=false (rather than panicking) when given no stores. This is the
+// empty-set contract relied on by getMeans and by callers that may receive a
+// constraint conjunction matching no stores.
+func TestComputeMeansForStoreSetEmpty(t *testing.T) {
+	loadProvider := &testLoadInfoProvider{
+		t:      t,
+		sloads: map[roachpb.StoreID]storeLoadAndNodeID{},
+		nloads: map[roachpb.NodeID]*NodeLoad{},
+	}
+	scratchNodes := map[roachpb.NodeID]*NodeLoad{}
+	scratchStores := map[roachpb.StoreID]struct{}{}
+
+	for _, name := range []string{"nil slice", "empty slice"} {
+		t.Run(name, func(t *testing.T) {
+			var stores []roachpb.StoreID
+			if name == "empty slice" {
+				stores = []roachpb.StoreID{}
+			}
+			means, ok := computeMeansForStoreSet(loadProvider, stores, scratchNodes, scratchStores)
+			require.False(t, ok)
+			require.Equal(t, meansLoad{}, means)
+		})
+	}
 }

--- a/pkg/kv/kvserver/allocator/mmaprototype/rebalance_advisor.go
+++ b/pkg/kv/kvserver/allocator/mmaprototype/rebalance_advisor.go
@@ -87,7 +87,15 @@ func (a *allocatorState) BuildMMARebalanceAdvisor(
 	scratchNodes := map[roachpb.NodeID]*NodeLoad{}
 	scratchStores := map[roachpb.StoreID]struct{}{}
 	cands = append(cands, existing)
-	means := computeMeansForStoreSet(a.cs, cands, scratchNodes, scratchStores)
+	means, ok := computeMeansForStoreSet(a.cs, cands, scratchNodes, scratchStores)
+	if !ok {
+		// Unreachable: cands always contains at least `existing`. Assert in
+		// test builds; in production, fall back to a no-op advisor rather than
+		// return a zero-valued means that would misclassify stores. Gating on
+		// !ok avoids variadic arg boxing on the success path.
+		assertTruef(context.Background(), false, "computeMeansForStoreSet returned !ok for non-empty cands=%v", cands)
+		return NoopMMARebalanceAdvisor()
+	}
 	return &MMARebalanceAdvisor{
 		existingStoreID: existing,
 		means:           means,

--- a/pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/empty_store_set.txt
+++ b/pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/empty_store_set.txt
@@ -1,0 +1,12 @@
+# Regression test for #170171 / #170136: a store-leaseholder message that
+# arrives while the MMA allocator has no stores registered must not panic
+# ("no stores for meansForStoreSet: []").
+#
+# This case is reachable in production at process startup, before any
+# gossip callback has called Allocator.SetStore. With the fix, getMeans
+# returns ok=false and processStoreLeaseholderMsgInternal returns early.
+
+# No set-store: MMA has zero stores registered.
+store-leaseholder-msg
+store-id=1
+----

--- a/pkg/kv/kvserver/allocator/mmaprototype/testdata/means_memo
+++ b/pkg/kv/kvserver/allocator/mmaprototype/testdata/means_memo
@@ -77,3 +77,9 @@ stores: 1, 2, 3
 store-means (load,cap,util): cpu: (43, unknown, 0.00) write-bw: (166, 566, 0.29) bytes: (1800, 4533, 0.40)
    secondary-load: [100 0]
 node-mean cpu (load,cap,util): (225, 750, 0.30)
+
+# Constraint that no store satisfies returns ok=false rather than panicking.
+get-means
++=purple
+----
+ok: false (no stores match)

